src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java

```diff
@@ -187,20 +190,33 @@ private CompletableFuture<Result<Request>> fetchItemAndLoanForPageTlrReplacement
       .toCompletableFuture();
   }
 
-  private Result<Request> findItemForPageTlr(Request request) {
+  private CompletableFuture<Result<Request>> findItemForPageTlr(Request request) {
     return getFirstAvailableItem(request)
-      .map(request::withItem)
-      .map(Result::succeeded)
-      .orElseGet(() -> failedValidation(
+      .thenApply(item -> item
+        .map(request::withItem)
+        .map(Result::succeeded)
+        .orElseGet(() -> failedValidation(
         "Cannot create page TLR for this instance ID - no available items found", INSTANCE_ID,
-        request.getInstanceId()));
+          request.getInstanceId())));
   }
 
-  private Optional<Item> getFirstAvailableItem(Request request) {
+  private CompletableFuture<Optional<Item>> getFirstAvailableItem(Request request) {
+    List<Item> availableItems = getAvailableItems(request).collect(toList());
+    return availableItems.size() == 1
+      ? completedFuture(Optional.of(availableItems.get(0)))
+      : completedFuture(availableItems)
+      .thenCombine(locationRepository.fetchLocationsByServicePointId(request.getPickupServicePointId()),
+        (items, listResult) -> items.stream()
+          .filter(item -> listResult.value().stream()
+            .anyMatch(location -> location.getId().equals(item.getLocationId())))
+          .findFirst()
+          .or(() -> getAvailableItems(request).findFirst()));
```

Comment:
Please make this more readable. Please find a way to avoid returning `Optional` wrapped in `CompletableFuture`, I think we should return `null` in such cases instead of empty `Optional`. And consider the case when list of available items is empty.